unified_planning/io/ma_pddl_writer.py

```diff
@@ -0,0 +1,1066 @@
+        """
+        This function simplifies an FNode to convert it into an FNode of type Dot and substitute the new FNode for the original expression.
+        """
+        # Initialize a simplifier object
+        o = Simplifier(self.problem.env)
+        # Define a helper function to simplify argument
+        def simplify_arg(arg):
+            # Simplify the argument
+            arg = o.simplify(arg)
+            if arg.is_not():
+                arg = self.problem.env.expression_manager.Not(arg)
+            return arg
+
+        # Initialize an empty list to store simplified arguments
+        args_list = []
+        # Iterate through all arguments of the node
+        for arg in node.args:
+            if node.is_fluent_exp():
+                # If the fluent is in the agent's fluents
+                if node.fluent() in agent.fluents:
+                    # Create a new dot expression
+                    new_dot_expression = self.problem.env.expression_manager.Dot(
+                        agent, node
+                    )
+                    return new_dot_expression
+            elif len(arg.args) > 1:
+                # Extend args_list with both sub-arguments simplified
+                args_list.extend([simplify_arg(arg.args[0]), simplify_arg(arg.args[1])])
+            else:
+                # Append arg to args_list after simplify
+                args_list.append(simplify_arg(arg))
+
+        # Create a new dot expression from the list of simplified args
+        new_dot_expression = self.sub_exp_dot(args_list, agent, node)
+        return new_dot_expression
+
+    def sub_exp(
+        self,
+        old_exp: "up.model.FNode",
+        new_exp: "up.model.FNode",
+        expression: "up.model.FNode",
+    ):
+        subs_map: Dict[Expression, Expression] = {}
+        sub = Substituter(self.problem.env)
+        subs_map[old_exp] = new_exp
+        new_expression = sub.substitute(expression, subs_map)
+        return new_expression
+
+    def sub_exp_dot(
+        self,
+        list_old_exp: List["up.model.FNode"],
+        agent: "up.model.multi_agent.Agent",
+        expression,
+    ):
+        subs_map: Dict[Expression, Expression] = {}
+        sub = Substituter(self.problem.env)
+        for old_exp in list_old_exp:
+            if old_exp.fluent() in agent.fluents:
+                arg_dot = self.problem.env.expression_manager.Dot(agent, old_exp)
+                subs_map[old_exp] = arg_dot
+            else:
+                return old_exp
+        new_expression = sub.substitute(expression, subs_map)
+        return new_expression
+
+    def add_agent_prefix(self, string: str, fluents_list: List["Fluent"]):
+        match = re.search(r"\((\w+)\s\?", string)
+        if match:
+            found = match.group(1)
+        else:
+            return string
+        for obj in fluents_list:
+            if obj.name == found:
+                return re.sub(r"\((\w+)\s\?", r"(a_\1 ?", string)
+        return string
```

Comment:
1. These methods are very strange for a printer: why do you need to build new expressions? I think it would be MUCH simpler to add an optional `agent` parameter to the `ConverterToMAPDDLString` and add the agent name where needed. For example:
```python
class ConverterToMAPDDLString(ConverterToPDDLString):
    """Expression converter to a MA-PDDL string."""

    def __init__(
        self,
        env: "up.environment.Environment",
        get_mangled_name: Callable[
            [
                Union[
                    "up.model.Type",
                    "up.model.Action",
                    "up.model.Fluent",
                    "up.model.Object",
                    "up.model.multi_agent.Agent",
                ]
            ],
            str,
        ],
        agent : Optional["up.model.multi_agent.Agent"]
    ):
        ConverterToPDDLString.__init__(self, env, get_mangled_name)
        self.agent = agent

    def walk_dot(self, expression, args):
        agent = expression.agent()
        fluent = expression.args[0].fluent()
        objects = expression.args[0].args
        return f'({self.get_mangled_name(fluent)} {self.get_mangled_name(agent)} {" ".join([self.convert(obj) for obj in objects])})'

    def walk_fluent_exp(self, expression, args):
        fluent = expression.fluent()
        return f'({self.get_mangled_name(fluent)} {self.get_mangled_name(self.agent) if self.agent is not None and fluent in self.agent.fluents else ""}{" " if len(args) > 0 else ""}{" ".join(args)})'
```
In this way, you can simply create different `ConverterToMAPDDLString` objects for different agents and use them. This should greatly simplify the printing, because you simply call the appropriate printer on a problem expression, without creating new FNodes.

2. these functions seems seriously buggy: why adding a `Not` if an expression is a `Not`? What happens if an expression has more than two arguments?  `sub_exp` is never used....

Reply:
The simplify() method aims to simplify the expression but does not take into account the "Not" operator. This is why I check if the FNode is of type "Not" and if it is, I remove it using `self.problem.env.expression_manager.Not(arg)`. This function will remove the "Not" operator if it is present, otherwise, it adds it.
I agree with your suggestion to use `ConverterToMAPDDLString` as it simplifies the printing process.
Done!